src/tasks.jl

```diff
@@ -0,0 +1,233 @@
+
+
+    res = Trace{typeof(vi),typeof(taskinfo)}(copy_vi(vi), Libtask.copy(task), copy(taskinfo));
+    # CTask(()->f());
+    if res.task.storage === nothing
+        res.task.storage = IdDict()
+    end
+    res.task.storage[:turing_trace] = res # create a backward reference in task_local_storage
+    return res
+end
+
+
+
+# step to the next observe statement, return log likelihood
+Libtask.consume(t::Trace) = (t.vi.num_produce += 1; consume(t.task))
+
+# Task copying version of fork for Trace.
+function fork(trace :: Trace, copy_vi::Function, is_ref :: Bool = false, set_retained_vns_del_by_spl!::Union{Function,Nothing} = nothing)
+    newtrace = copy(trace,copy_vi)
+    if is_ref
+        @assert set_retained_vns_del_by_spl! != nothing "[AdvancedPF] set_retained_vns_del_by_spl! is not set."
+        set_retained_vns_del_by_spl!(newtrace.vi)
+    end
+    newtrace.task.storage[:turing_trace] = newtrace
+    return newtrace
+end
+
+# PG requires keeping all randomness for the reference particle
+# Create new task and copy randomness
+function forkr(trace :: Trace, copy_vi::Function)
+
+    newtrace = Trace(trace.vi,trace.task ,trace.taskinfo,copy_vi)
+    newtrace.vi.num_produce = 0
+    return newtrace
+end
+
+current_trace() = current_task().storage[:turing_trace]
+
+const Particle = Trace
+
+"""
+Data structure for particle filters
+- effectiveSampleSize(pc :: ParticleContainer)
+- normalise!(pc::ParticleContainer)
+- consume(pc::ParticleContainer): return incremental likelihood
+"""
+mutable struct ParticleContainer{Tvi,TInfo} <: AbstractParticleContainer where {Tvi, TInfo <: AbstractTaskInfo}
+    vals::Vector{Trace{Tvi,TInfo}}
+    # A named tuple with functions to manipulate the particles vi.
+    manipulators::Dict{String,Function}
+    # logarithmic weights (Trace) or incremental log-likelihoods (ParticleContainer)
+    logWs::Vector{Float64}
+    #This corresponds to log p(x_{0:t}), therefore the log likelihood of the transitions up to this point.
+    #Especially important for ancestor sampling.
+    logpseq::Vector{Float64}
+    # log model evidence
+    logE::Float64
+    # helpful for rejuvenation steps, e.g. in SMC2
+    n_consume::Int
+
+
+end
+
+## Empty initilaizaiton
+ParticleContainer{Tvi,TInfo}() where {Tvi, TInfo <: AbstractTaskInfo} = ParticleContainer{Tvi,TInfo}(Dict{String,Function}(),0)
+ParticleContainer{Tvi,TInfo}(manipulators::Dict{String,Function}) where {Tvi, TInfo <: AbstractTaskInfo} = ParticleContainer{Tvi,TInfo}(manipulators,0)
+#Some initilaizaiton...
+function ParticleContainer{Tvi,TInfo}(manipulators::Dict{String,Function}, n::Int) where {Tvi, TInfo <: AbstractTaskInfo}
+    if !("copy" in keys(manipulators)) manipulators["copy"] = deepcopy end
+    if !("set_retained_vns_del_by_spl!" in keys(manipulators)) manipulators["set_retained_vns_del_by_spl!"] = (x) -> () end
+    ParticleContainer{Tvi,TInfo}( Vector{Trace{Tvi,TInfo}}(undef, n), manipulators, Float64[],Float64[], 0.0, 0)
+
+end
```

Comment:
In the current implementation in Turing there's no need for any special constructors, in particular not for constructors without particles. All particles could be created first, and then a container could be constructed with them, filling in zeros and other default values for the remaining fields. Did that change?

Reply:
No, they do not use it, however, my goal is to hide the Trace struct form Turing. Creating an empty particle container and filling it using push allows us to change the structure of the trace struct without the need for adapting the Turing code. This might be important when changing the consumer/producer setting to the more recent channels https://github.com/TuringLang/Turing.jl/issues/466.

Reply:
I think, this question does not depend on the implementation in Turing? The current version of Turing doesn't use any empty initialization anymore. No matter how particles are implemented and which structs are used, I guess usually one create an array of particles first if there are multiple particles (so there's not really any need for the `push!` function that adds `N` particles). If at some points particle have to be added dynamically (which is not the case in the current implementation in Turing), the `push!(::ParticleContainer, ::Particle)` function should be sufficient.

Reply:
I see, thanks for the comment. I will change it.